crates/bevy_ecs/src/storage/blob_array.rs

```diff
@@ -0,0 +1,455 @@
+                let layout =
+                    array_layout(&self.item_layout, cap).expect("array layout should be valid");
+                std::alloc::dealloc(self.data.as_ptr().cast(), layout);
+            }
+        }
+    }
+
+    /// Drops the last element in this [`BlobArray`].
+    ///
+    /// # Safety
+    // - `last_element_index` must correspond to the last element in the array.
+    // - After this method is called, the last element must not be used
+    // unless [`Self::initialize_unchecked`] is called to set the value of the last element.
+    pub unsafe fn drop_last_element(&mut self, last_element_index: usize) {
+        if let Some(drop) = self.drop {
+            // We set `self.drop` to `None` before dropping elements for unwind safety. This ensures we don't
+            // accidentally drop elements twice in the event of a drop impl panicking.
+            self.drop = None;
+            // SAFETY:
+            let item = self.get_unchecked_mut(last_element_index).promote();
+            // SAFETY:
+            unsafe { drop(item) };
+            self.drop = Some(drop);
+        }
+    }
+
+    /// Allocate a block of memory for the array. This should be used to initialize the array, do not use this
+    /// method if there are already elements stored in the array - use [`Self::realloc`] instead.
+    pub(super) fn alloc(&mut self, capacity: NonZeroUsize) {
+        if !self.is_zst() {
+            let new_layout = array_layout(&self.item_layout, capacity.get())
+                .expect("array layout should be valid");
+            let new_data
+            // SAFETY:
+            // - layout has non-zero size because capacity > 0, and the blob isn't ZST (`self.is_zst` == false)
+            = unsafe {std::alloc::alloc(new_layout)};
+
+            self.data = NonNull::new(new_data).unwrap_or_else(|| handle_alloc_error(new_layout));
+        }
+    }
+
+    /// Reallocate memory for this array.
+    /// For example, if the length (number of stored elements) reached the capacity (number of elements the current allocation can store),
+    /// you might want to use this method to increase the allocation, so more data can be stored in the array.
+    ///
+    /// # Safety
+    /// - `current_capacity` + `increment` doesn't overflow `usize`
+    /// - The size of the resulting array does not overflow `usize` (specifically, see the safety requirements of [`array_layout_unchecked`])
+    /// - `current_capacity` is indeed the current capacity of this array.
+    /// - After calling this method, the caller must update their saved capacity to reflect the change.
+    pub(super) unsafe fn realloc(
+        &mut self,
+        current_capacity: NonZeroUsize,
+        new_capacity: NonZeroUsize,
+    ) {
+        if !self.is_zst() {
+            let new_layout =
+            // SAFETY: Safety requirement 2
+            unsafe {
+                array_layout_unchecked(&self.item_layout, new_capacity.get())
+            };
```

Comment:
Same here.

Reply:
I think CI demands the safety comment be directly above the unsafe block, and if the let statement gets cut in the middle and indented because its too long the CI may not detect the safety comment, but I'll check